specs/crux/beacon-chain.md

```diff
@@ -0,0 +1,247 @@
+    for_ops(body.voluntary_exits, process_voluntary_exit)
+    for_ops(body.delegations, process_delegation)   # [New in Crux]
+```
+
+#### Delegation
+
+##### `get_validator_from_delegation`
+
+```python
+def get_validator_from_delegation_message(message: DelegationMessage, amount: Gwei) -> Validator:
+    effective_balance = min(amount - amount % EFFECTIVE_BALANCE_INCREMENT, MAX_EFFECTIVE_BALANCE)
+    return Validator(
+        pubkey=message.pubkey,
+        withdrawal_credentials=message.withdrawal_credentials,
+        activation_eligibility_epoch=FAR_FUTURE_EPOCH,
+        activation_epoch=FAR_FUTURE_EPOCH,
+        exit_epoch=FAR_FUTURE_EPOCH,
+        withdrawable_epoch=FAR_FUTURE_EPOCH,
+        effective_balance=effective_balance,
+    )
+```
+
+##### `process_delegation`
+
+```python
+def process_delegation(state: BeaconState, delegation: Delegation) -> None:
+    message = delegation.message
+    current_epoch = get_current_epoch(state)
+    assert  message.epoch <= current.epoch <= message.epoch + MAX_DELEGATION_INCLUSION_DELAY
+
+    delegating_index = message.delegating_index
+    min_delegating_balance = MAX_EFFECTIVE_BALANCE + MIN_DEPOSIT_AMOUNT + DELEGATION_TRANSACTION_COST
+    assert state.balances[delegating_index] >= min_delegating_balance
+    amount = state.balances[delegating_index] - MAX_EFFECTIVE_BALANCE - DELEGATION_TRANSACTION_COST
+
+    delegating_validator = state.validators[delegating_index]
+    assert is_active_validator(delegating_validator, current_epoch)
+    assert not delegating_validator.slashed 
+
+    assert delegating_validator.delegate == delegating_index
+
+    delegating_pubkey = message.delegating_pubkey
+    assert hash(delegating_pubkey)[1:] == delegating_validator.withdrawal_credentials[1:]
+
+    domain = compute_domain(DOMAIN_DELEGATION)
+    signing_root = compute_signing_root(message, domain)
+    assert bls.Verify(delegating_pubkey, signing_root, delegation.signature)
+
+    validator_pubkeys = [v.pubkey for v in state.validators]
+    if pubkey not in validator_pubkeys:
+        index = len(state.validators)
+        state.validators.append(get_validator_from_delegation(message, amount))
+    else:
+        index = ValidatorIndex(validator_pubkeys.index(pubkey))
+        assert index > delegating_index
+        assert state.validators[index].exit_epoch == FAR_FUTURE_EPOCH
+
+    delegating_validator.delegate = index
+    proposer_index = get_beacon_proposer_index(state)
+    decrease_balance(state, delegating_validator, amount + DELEGATION_TRANSACTION_COST)
+    increase_balance(state, index, amount)
+    increase_balance(state, proposer_index, DELEGATION_TRANSACTION_COST)
+```
+
+### Epoch processing
+
+```python
+def process_epoch(state: BeaconState) -> None:
+    process_justification_and_finalization(state)
+    process_inactivity_updates(state)
+    process_rewards_and_penalties(state)
+    process_delegation_transfers(state)  # [New in Crux]
```

Comment:
Should this be done after processing any other items that may modify balances?

Reply:
I couldn't think of any danger on doing this here. I think it has to be before `process_registry_updates`  and `process_effective_balance` though.